File: server/src/fetch.ts

const DIGITEC_FETCH_URL = 'https://www.digitec.ch/api/graphql/get-social-shoppings';

const HEADERS = {
	'Accept': '*/*',
	'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.6 Safari/605.1.1',
	'Content-Type': 'application/json',
	'Origin': 'https://www.digitec.ch',
	'Referer': 'https://www.digitec.ch/'
};

const GRAPHQL_BODY = [{
	operationName: 'GET_SOCIAL_SHOPPINGS',
	query: `query GET_SOCIAL_SHOPPINGS($take: Int, $latest: String) {
        socialShopping(take: $take, latest: $latest) {
          latestTransactionTimeStamp
          items {
            id
            userName
            cityName
            dateTime
            imageUrl
            brandName
            fullProductName
            displayPrice {
              amountInclusive
              amountExclusive
              currency
            }
            oAuthProviderName
            targetUserName
            quote
            voteTypeId
            productTypeName
            socialShoppingTransactionTypeId
            url
            rating
            searchString
          }
        }
      }`,
	variables: { 'take': 6, 'latest': null }
}];

const BODY_CONTENT = JSON.stringify(GRAPHQL_BODY);

let lastFetchedItemIds = [];


export const fetchFeedItems = async () => {
	const response = await fetch(DIGITEC_FETCH_URL, {
		method: 'POST',
		headers: HEADERS,
		body: BODY_CONTENT
	});

	if (!response.ok) {
    const errorText = await response.text();
    throw new Error(`Failed to fetch feed items: ${response.status} ${response.statusText}. Response: ${errorText}`);
  }

	const data = await response.json();
	const currentItems = data[0].data.socialShopping.items;

	const currentItemIds = currentItems.map(generateItemIdentifier);
	const newIds = currentItemIds.filter((id) => !lastFetchedItemIds.includes(id));

	const newItems = currentItems.filter((item) => {
		const itemIdentifier = generateItemIdentifier(item);
		return newIds.includes(itemIdentifier);
	});

	lastFetchedItemIds = [...newIds, ...lastFetchedItemIds].slice(0, 50);
	return newItems;
};


function generateItemIdentifier(item) {
	return `${item.userName}_${item.dateTime}_${item.socialShoppingTransactionTypeId}_${item.url}`;
}

File: go-new-backend/fetch.go
package main

import (
	"bytes"
	"encoding/json"
	"io"
	"log"
	"net/http"
	"strings"
)

const digitecFetchUrl = "https://www.digitec.ch/api/graphql/get-social-shoppings"
const userAgent = "Mozilla/5.0 (Linux; Android 10; K) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/134.0.0.0 Safari/537.36"

var headers = map[string]string{
	"accept":         "*/*",
	"user-agent":     userAgent,
	"content-type":   "application/json",
	"origin":         "https://www.digitec.ch",
	"referer":        "https://www.digitec.ch/en/daily-deal",
	"sec-fetch-mode": "cors",
}

var graphqlBody = GraphQLRequest{
	OperationName: "GET_SOCIAL_SHOPPINGS",
	Query: `query GET_SOCIAL_SHOPPINGS($take: Int, $latest: String) {
	socialShopping(take: $take, latest: $latest) {
	  latestTransactionTimeStamp
	  items {
		id
		userName
		cityName
		dateTime
		imageUrl
		brandName
		fullProductName
		displayPrice {
		  amountInclusive
		  amountExclusive
		  currency
		}
		oAuthProviderName
		targetUserName
		quote
		voteTypeId
		productTypeName
		socialShoppingTransactionTypeId
		url
		rating
		searchString
	  }
	}
  }`,
	Variables: map[string]interface{}{
		"take":   10,
		"latest": nil,
	},
}

func FetchItems() []FeedItem {
	bodyContent, err := json.Marshal(graphqlBody)
	if err != nil {
		log.Fatal("Error encoding JSON:", err)
	}

	req, err := http.NewRequest("POST", digitecFetchUrl, bytes.NewBuffer(bodyContent))
	if err != nil {
		log.Fatal("Error creating request:", err)
	}

	for key, value := range headers {
		req.Header.Set(key, value)
	}

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		log.Fatal("Error making request:", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		log.Fatal("Error reading response:", err)
	}

	var graphqlResponse GraphQLResponse
	err = json.Unmarshal(body, &graphqlResponse)
	if err != nil {
		log.Fatal("Error decoding JSON:", err)
	}

	items := graphqlResponse.Data.SocialShopping.Items
	for i := range items {
		items[i].FullProductName = strings.TrimSpace(items[i].FullProductName)
	}

	return items
}
